File: mobile/src/main/java/net/mm2d/dmsexplorer/view/delegate/ContentListActivityDelegate.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.delegate;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.KeyEvent;
import android.view.View;

import net.mm2d.android.util.ViewUtils;
import net.mm2d.dmsexplorer.R;
import net.mm2d.dmsexplorer.Repository;
import net.mm2d.dmsexplorer.databinding.ContentListActivityBinding;
import net.mm2d.dmsexplorer.settings.Settings;
import net.mm2d.dmsexplorer.view.base.BaseActivity;
import net.mm2d.dmsexplorer.view.dialog.DeleteDialog.OnDeleteListener;
import net.mm2d.dmsexplorer.viewmodel.ContentListActivityModel;
import net.mm2d.dmsexplorer.viewmodel.ContentListActivityModel.CdsSelectListener;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public abstract class ContentListActivityDelegate implements CdsSelectListener, OnDeleteListener {
    private static final String KEY_SCROLL_POSITION = "KEY_SCROLL_POSITION";
    private static final String KEY_SCROLL_OFFSET = "KEY_SCROLL_OFFSET";

    public static ContentListActivityDelegate create(@NonNull final BaseActivity activity) {
        final ContentListActivityBinding binding = DataBindingUtil.setContentView(activity, R.layout.content_list_activity);
        if (binding.cdsDetailContainer == null) {
            return new ContentListActivityDelegateOnePane(activity, binding);
        }
        return new ContentListActivityDelegateTwoPane(activity, binding);
    }

    private final BaseActivity mActivity;
    private final ContentListActivityBinding mBinding;
    private ContentListActivityModel mModel;

    ContentListActivityDelegate(
            @NonNull final BaseActivity activity,
            @NonNull final ContentListActivityBinding binding) {
        mActivity = activity;
        mBinding = binding;
    }

    @NonNull
    protected BaseActivity getActivity() {
        return mActivity;
    }

    @NonNull
    public ContentListActivityBinding getBinding() {
        return mBinding;
    }

    @Nullable
    protected ContentListActivityModel getModel() {
        return mModel;
    }

    protected abstract boolean isTwoPane();

    @CallSuper
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        final BaseActivity activity = getActivity();
        final Repository repository = Repository.get();
        try {
            mModel = new ContentListActivityModel(getActivity(), repository, this, isTwoPane());
        } catch (final IllegalStateException ignored) {
            activity.finish();
            return;
        }
        final ContentListActivityBinding binding = getBinding();

        binding.toolbar.setPopupTheme(new Settings(activity).getThemeParams().getPopupThemeId());
        binding.setModel(mModel);
        activity.setSupportActionBar(mBinding.toolbar);
        activity.getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        if (savedInstanceState != null) {
            restoreScroll(savedInstanceState);
        }
        repository.getThemeModel().setThemeColor(getActivity(), mModel.toolbarBackground, 0);
    }

    @CallSuper
    public void onSaveInstanceState(@NonNull final Bundle outState) {
        if (mModel == null) {
            return;
        }
        saveScroll(outState);
    }

    @CallSuper
    public void onStart() {
        if (mModel == null) {
            return;
        }
        mModel.syncSelectedEntity();
    }

    private void restoreScroll(@NonNull final Bundle savedInstanceState) {
        final int position = savedInstanceState.getInt(KEY_SCROLL_POSITION, 0);
        final int offset = savedInstanceState.getInt(KEY_SCROLL_OFFSET, 0);
        if (position == 0 && offset == 0) {
            return;
        }
        final RecyclerView recyclerView = mBinding.recyclerView;
        ViewUtils.execOnLayout(recyclerView, () -> {
            recyclerView.scrollToPosition(position);
            recyclerView.post(() -> recyclerView.scrollBy(0, offset));
        });
    }

    private void saveScroll(@NonNull final Bundle outState) {
        final RecyclerView recyclerView = mBinding.recyclerView;
        if (recyclerView.getChildCount() == 0) {
            return;
        }
        final View view = recyclerView.getChildAt(0);
        outState.putInt(KEY_SCROLL_POSITION, recyclerView.getChildAdapterPosition(view));
        outState.putInt(KEY_SCROLL_OFFSET, -view.getTop());
    }

    public boolean onBackPressed() {
        return mModel != null && mModel.onBackPressed();
    }

    public boolean onKeyLongPress(
            final int keyCode,
            final KeyEvent event) {
        if (mModel != null && keyCode == KeyEvent.KEYCODE_BACK) {
            mModel.terminate();
            return true;
        }
        return false;
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/view/delegate/ContentListActivityDelegate.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.delegate

import android.os.Bundle
import android.view.KeyEvent
import android.view.MenuItem
import androidx.annotation.CallSuper
import androidx.core.view.doOnLayout
import androidx.core.view.updatePaddingRelative
import net.mm2d.dmsexplorer.R
import net.mm2d.dmsexplorer.Repository
import net.mm2d.dmsexplorer.databinding.ContentListActivityBinding
import net.mm2d.dmsexplorer.settings.Settings
import net.mm2d.dmsexplorer.util.observe
import net.mm2d.dmsexplorer.view.base.BaseActivity
import net.mm2d.dmsexplorer.view.dialog.DeleteDialog.OnDeleteListener
import net.mm2d.dmsexplorer.viewmodel.ContentListActivityModel
import net.mm2d.dmsexplorer.viewmodel.ContentListActivityModel.CdsSelectListener
import net.mm2d.dmsexplorer.viewmodel.adapter.RecyclerViewBindingAdapter

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
abstract class ContentListActivityDelegate internal constructor(
    protected val activity: BaseActivity,
    val binding: ContentListActivityBinding,
) : CdsSelectListener, OnDeleteListener {
    protected var model: ContentListActivityModel? = null
        private set

    protected abstract val isTwoPane: Boolean

    @CallSuper
    fun onCreate(savedInstanceState: Bundle?) {
        val activity = activity
        val repository = Repository.get()
        val model = try {
            ContentListActivityModel(activity, repository, this, isTwoPane)
        } catch (ignored: IllegalStateException) {
            activity.finish()
            return
        }
        this.model = model
        binding.toolbar.popupTheme = Settings.get().themeParams.popupThemeId
        binding.toolbar.setBackgroundColor(model.toolbarBackground)
        binding.toolbar.title = model.title
        model.getSubtitleFlow().observe(activity) { binding.toolbar.subtitle = it }
        binding.swipeRefreshLayout.setColorSchemeColors(*model.refreshColors)
        binding.swipeRefreshLayout.setDistanceToTriggerSync(model.distanceToTriggerSync)
        binding.swipeRefreshLayout.setOnRefreshListener(model.onRefreshListener)
        binding.swipeRefreshLayout.setProgressBackgroundColorSchemeColor(model.progressBackground)
        model.getIsRefreshingFlow().observe(activity) { binding.swipeRefreshLayout.isRefreshing = it }

        binding.recyclerView.itemAnimator = model.itemAnimator
        binding.recyclerView.layoutManager = model.cdsListLayoutManager
        binding.recyclerView.adapter = model.contentListAdapter
        val padding = activity.resources.getDimensionPixelSize(
            if (model.focusable) R.dimen.list_scale_padding else R.dimen.list_non_scale_padding,
        )

        if (binding.cdsDetailContainer != null) {
            binding.recyclerView.updatePaddingRelative(start = padding)
        } else {
            binding.recyclerView.updatePaddingRelative(start = padding, end = padding)
        }

        model.getScrollPositionFlow().observe(activity) {
            RecyclerViewBindingAdapter.setScrollPosition(binding.recyclerView, it)
        }
        activity.setSupportActionBar(binding.toolbar)
        activity.supportActionBar?.setDisplayHomeAsUpEnabled(true)
        savedInstanceState?.let { restoreScroll(it) }
        repository.themeModel.setThemeColor(activity, model.toolbarBackground, 0)
    }

    @CallSuper
    fun onSaveInstanceState(outState: Bundle) {
        model ?: return
        saveScroll(outState)
    }

    @CallSuper
    open fun onStart() {
        model?.syncSelectedEntity()
    }

    private fun restoreScroll(savedInstanceState: Bundle) {
        val position = savedInstanceState.getInt(KEY_SCROLL_POSITION, 0)
        val offset = savedInstanceState.getInt(KEY_SCROLL_OFFSET, 0)
        if (position == 0 && offset == 0) {
            return
        }
        val recyclerView = binding.recyclerView
        recyclerView.doOnLayout {
            recyclerView.scrollToPosition(position)
            recyclerView.post { recyclerView.scrollBy(0, offset) }
        }
    }

    private fun saveScroll(outState: Bundle) {
        val recyclerView = binding.recyclerView
        if (recyclerView.childCount == 0) {
            return
        }
        val view = recyclerView.getChildAt(0)
        outState.putInt(KEY_SCROLL_POSITION, recyclerView.getChildAdapterPosition(view))
        outState.putInt(KEY_SCROLL_OFFSET, -view.top)
    }

    fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_sort) {
            model?.onSortMenuClicked()
            return true
        }
        return false
    }

    fun onBackPressed(): Boolean {
        return model?.onBackPressed() == true
    }

    fun onKeyLongPress(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            model?.terminate()
            return true
        }
        return false
    }

    companion object {
        private const val KEY_SCROLL_POSITION = "KEY_SCROLL_POSITION"
        private const val KEY_SCROLL_OFFSET = "KEY_SCROLL_OFFSET"

        fun create(activity: BaseActivity): ContentListActivityDelegate {
            val binding = ContentListActivityBinding.inflate(activity.layoutInflater)
            activity.setContentView(binding.root)

            return if (binding.cdsDetailContainer == null) {
                ContentListActivityDelegateOnePane(activity, binding)
            } else {
                ContentListActivityDelegateTwoPane(
                    activity,
                    binding,
                )
            }
        }
    }
}
